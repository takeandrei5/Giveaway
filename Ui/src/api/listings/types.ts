import { PaginatedResult } from '../../utils/types';

export type ItemData = { id: string; image: string; title: string; createdAt: Date };

export type CreateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export type UpdateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export interface FetchListingsResponse {
	listings: PaginatedResult<ItemData>;
}

export type ListingInformation = {
	id: string;
	title: string;
	description: string;
	category: number;
	createdAt: Date;
	images: string[];
};

export type OwnerInformation = {
	email: string;
	name: string;
	image: string;
};

export type FetchListingDetailsResponse = { listingInfo: ListingInformation } & { ownerInfo: OwnerInformation };
