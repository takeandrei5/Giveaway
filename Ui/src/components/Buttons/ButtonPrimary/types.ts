import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = Omit<ButtonProps, 'color'>;
