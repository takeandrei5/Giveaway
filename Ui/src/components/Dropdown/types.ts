export type DropdownProps = {
	id: string;
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
	name: string;
	value: string | number | readonly string[];
	isInvalid?: false;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
