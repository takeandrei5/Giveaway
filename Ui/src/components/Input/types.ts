export type InputProps = {
	id: string;
	name: string;
	placeholder: string;
	value?: string;
	onBlur?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onChange?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyUp?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	disabled?: boolean;
	multiline?: boolean;
	isInvalid?: boolean;
	height?: string | number;
	label?: string;
	leftIcon?: JSX.Element;
	rightIcon?: JSX.Element;
	rows?: number;
	width?: string | number;
};
