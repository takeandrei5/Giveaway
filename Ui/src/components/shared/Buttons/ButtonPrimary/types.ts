import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = ButtonProps;
