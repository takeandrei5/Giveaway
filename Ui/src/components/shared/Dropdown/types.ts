export type DropdownProps = {
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
