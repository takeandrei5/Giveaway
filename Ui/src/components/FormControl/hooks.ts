import { FieldMetaProps } from 'formik';
import { useEffect, useState } from 'react';

import { useCheckFormIsInvalidProps } from './types';

const useCheckFormIsInvalid = (meta: FieldMetaProps<unknown>): useCheckFormIsInvalidProps => {
	const [isInvalid, setIsInvalid] = useState(false);

	useEffect(() => {
		setIsInvalid(meta.touched && !!meta.error);
	}, [meta.touched, meta.error]);

	return { isInvalid };
};

export { useCheckFormIsInvalid };
