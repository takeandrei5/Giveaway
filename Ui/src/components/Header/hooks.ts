import { NextRouter, useRouter } from 'next/router';
import { useCallback } from 'react';

const useLogin = () => {
	const router: NextRouter = useRouter();

	const handleSignIn: () => Promise<boolean> = useCallback(
		async () => await router.replace('/api/auth/login'),
		[router]
	);
	const handleSignInWithReturnTo: (returnTo: string) => Promise<boolean> = useCallback(
		async (returnTo: string) => await router.replace(`/api/auth/login?returnTo=${returnTo}`),
		[router]
	);
	const handleSignOut: () => Promise<boolean> = useCallback(() => router.replace('/api/auth/logout'), [router]);

	return { handleSignIn, handleSignInWithReturnTo, handleSignOut };
};

export { useLogin };
