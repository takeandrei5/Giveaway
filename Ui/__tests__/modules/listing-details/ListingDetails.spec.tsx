import { useUser } from '@auth0/nextjs-auth0';
import { ListingDetails } from '@modules/listing-details';
import { useFetchListingDetails } from '@modules/listing-details/hooks';
import { ListingDetailsProps } from '@modules/listing-details/types';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useGetAccessToken } from '@utils/hooks';
import { useRouter } from 'next/router';
import { ListingInformation } from '../../../src/api/listings/types';

jest.mock('dateformat', () => jest.fn());

jest.mock('next/router', () => ({
	...jest.requireActual('next/router'),
	useRouter: jest.fn(),
}));

jest.mock('@utils/hooks', () => ({
	useGetAccessToken: jest.fn(),
}));

jest.mock('@modules/listing-details/hooks', () => ({
	useFetchListingDetails: jest.fn<any, [id: string, isAccessTokenLoaded: boolean]>(),
}));

jest.mock('@auth0/nextjs-auth0', () => ({ ...jest.requireActual('@auth0/nextjs-auth0'), useUser: jest.fn() }));

describe('ListingDetails', () => {
	const listingInfo: ListingInformation = {
		id: 'listing-id',
		title: 'listing-title',
		description: 'listing-description',
		category: 0,
		createdAt: new Date('2020-01-01'),
		images: ['https://via.placeholder.com/150'],
	};

	const accessTokenFetched: boolean = true;
	const deleteListingMutateMock = jest.fn();

	(useGetAccessToken as unknown as jest.Mock).mockImplementation(() => ({
		isFetched: accessTokenFetched,
		data: 'access-token',
	}));

	(useUser as unknown as jest.Mock).mockImplementation(() => ({ user: { name: 'John Doe' } }));

	(useFetchListingDetails as unknown as jest.Mock<any, [id: string, isAccessTokenLoaded: boolean]>).mockImplementation(
		(id: string, isAccessTokenFetched: boolean) => ({
			isLoading: false,
			listingInfo,
			ownerInfo: {},
			deleteListingMutate: deleteListingMutateMock,
		})
	);

	const props: ListingDetailsProps = {
		id: 'test-id',
	};

	it('should match snapshot', () => {
		// Act
		const { container } = render(<ListingDetails {...props} />);

		// Assert
		expect(container).toMatchSnapshot();
	});

	describe('when the buttons are clicked', () => {
		it('should redirect user to `/update-listing/${id}` page if the update listing button is clicked', async () => {
			// Act
			let route = '';

			(useRouter as unknown as jest.Mock).mockImplementation(() => ({
				route,
				push: jest.fn((value: string) => (route = value)),
			}));

			const { getByText } = render(<ListingDetails {...props} />);

			await userEvent.click(getByText('Update listing'));

			// Assert
			expect(route).toBe(`/update-listing/${props.id}`);
		});

		it('should execute `deleteListingMutate` if the delete listing button is clicked', async () => {
			// Act
			const { getByText } = render(<ListingDetails {...props} />);

			await userEvent.click(getByText('Delete listing'));

			// Assert
			expect(deleteListingMutateMock).toHaveBeenCalledTimes(1);
		});
	});
});
