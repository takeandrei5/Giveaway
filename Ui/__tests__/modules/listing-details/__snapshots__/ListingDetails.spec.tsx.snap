// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetails should match snapshot 1`] = `
<div>
  <div
    class="chakra-skeleton css-r69txw"
  >
    <div
      class="css-3a3hw0"
    >
      <span
        class="chakra-text css-f3be6h"
        data-testid="typography"
      >
        Details
      </span>
      <div
        class="css-k008qs"
      >
        <button
          class="chakra-button css-g5yezb"
          data-testid="button-primary"
          type="button"
        >
          <span
            class="chakra-button__icon css-1wh2kri"
          >
            <svg
              aria-hidden="true"
              color="white"
              fill="currentColor"
              focusable="false"
              height="1rem"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 24 24"
              width="1rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17.77 3.77L16 2 6 12l10 10 1.77-1.77L9.54 12z"
              />
            </svg>
          </span>
          <span
            class="chakra-text css-7fixei"
            data-testid="typography"
          >
            Back
          </span>
        </button>
        <div
          class="css-glbj5v"
        >
          <div
            class="chakra-skeleton css-cdkrf0"
          >
            <button
              class="chakra-button css-37dqwd"
              data-testid="button-primary"
              type="button"
            >
              <span
                class="chakra-button__icon css-1wh2kri"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  font-size="medium"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1.7507,16.0022 C3.3517,20.0982 7.3367,23.0002 11.9997,23.0002 C18.0747,23.0002 22.9997,18.0752 22.9997,12.0002 M22.2497,7.9982 C20.6487,3.9012 16.6627,1.0002 11.9997,1.0002 C5.9247,1.0002 0.9997,5.9252 0.9997,12.0002 M8.9997,16.0002 L0.9997,16.0002 L0.9997,24.0002 M22.9997,0.0002 L22.9997,8.0002 L14.9997,8.0002"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
              </span>
              <span
                class="chakra-text css-7fixei"
                data-testid="typography"
              >
                Update listing
              </span>
            </button>
          </div>
          <div
            class="chakra-skeleton css-cdkrf0"
          >
            <button
              class="chakra-button css-37dqwd"
              data-testid="button-primary"
              type="button"
            >
              <span
                class="chakra-button__icon css-1wh2kri"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  font-size="medium"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4,5 L20,5 L20,23 L4,23 L4,5 Z M1,5 L23,5 M9,1 L15,1 L15,5 L9,5 L9,1 Z M9,1 L15,1 L15,5 L9,5 L9,1 Z M15,9 L15,19 M9,9 L9,19"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
              </span>
              <span
                class="chakra-text css-7fixei"
                data-testid="typography"
              >
                Delete listing
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-19hwa6l"
    >
      <div
        class="slick-slider slick-initialized"
      >
        <div
          class="slick-list"
          style="height: 0px;"
        >
          <div
            class="slick-track"
            style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
          >
            <div
              aria-hidden="false"
              class="slick-slide slick-active slick-current"
              data-index="0"
              style="outline: none; width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="css-n8wih0"
                >
                  <span
                    style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                  >
                    <span
                      style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
                    />
                    <img
                      alt="image"
                      data-nimg="responsive"
                      decoding="async"
                      draggable="false"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="width: 0px; display: block; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
                      tabindex="-1"
                    />
                    <noscript />
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-h7npbh"
    >
      <div
        class="css-1b27p5s"
      >
        <div
          class="css-1jkzbrs"
        >
          <span
            class="chakra-text css-wwpj52"
            data-testid="typography"
          >
            listing-title
          </span>
          <span
            class="chakra-text css-1jv3994"
            data-testid="typography"
          >
            listing-description
          </span>
          <div
            class="css-8xl60i"
          >
            <span
              class="chakra-text css-1m60o1a"
              data-testid="typography"
            >
              Added on: 
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-1pip4vl"
      >
        <div
          class="css-rqk74a"
        >
          <span
            class="chakra-text css-84comp"
            data-testid="typography"
          >
            Contact Information
          </span>
          <span
            class="chakra-avatar css-2u22ig"
          >
            <svg
              aria-label=" avatar"
              class="chakra-avatar__svg css-16ite8i"
              role="img"
              viewBox="0 0 128 128"
            >
              <path
                d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                fill="currentColor"
              />
              <path
                d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                fill="currentColor"
              />
            </svg>
          </span>
          <div
            class="css-w28dba"
          >
            <span
              class="chakra-text css-1dioq7p"
              data-testid="typography"
            >
              Owner:
            </span>
            <span
              class="chakra-text css-lk79i8"
              data-testid="typography"
            />
          </div>
          <div
            class="css-w28dba"
          >
            <span
              class="chakra-text css-1dioq7p"
              data-testid="typography"
            >
              Email:
            </span>
            <span
              class="chakra-text css-lk79i8"
              data-testid="typography"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
